File: python/poisson.py
#!/usr/bin/env python3
import sys
import time
import numpy as np
import numba as nb

class Metrics:
    def __init__(self, nx, ny, nz):
        self.dx = 1.0 / (nx-1)
        self.dy = 1.0 / (ny-1)
        self.dz = 1.0 / (nz-1)
        self.ax = 1.0 / (self.dx*self.dx)
        self.ay = 1.0 / (self.dy*self.dy)
        self.az = 1.0 / (self.dz*self.dz)
        self.dxyz = 2.0 * (self.ax + self.ay + self.az)
        self.inv_dxyz = 1.0/self.dxyz

def simple_update(u0, u1, f, ax, ay, az, inv_dxyz):
    norm_du = 0.0
    for i in nb.prange(1, u0.shape[0]-1): # prange => parallel
      for j in range(1, u0.shape[1]-1):
        for k in range(1, u0.shape[2]-1):
          u1[i,j,k] = -f[i,j,k]
          u1[i,j,k] += ax*u0[i-1, j, k]
          u1[i,j,k] += ax*u0[i+1, j, k]
          u1[i,j,k] += ay*u0[i, j-1, k]
          u1[i,j,k] += ay*u0[i, j+1, k]
          u1[i,j,k] += az*u0[i, j, k-1]
          u1[i,j,k] += az*u0[i, j, k+1]
          u1[i,j,k] *= inv_dxyz
          du = u1[i,j,k] - u0[i,j,k]
          norm_du += du*du
    return np.sqrt(norm_du)

def vectorized_update(u0, u1, f, ax, ay, az, inv_dxyz):
    ni, nj, nk = u0.shape
    i = slice(1,ni-1)
    j = slice(1,nj-1)
    k = slice(1,nk-1)
    u1[i,j,k] = -f[i,j,k]
    u1[i,j,k] += ax*u0[:ni-2,j,k]
    u1[i,j,k] += ax*u0[2:,j,k]
    u1[i,j,k] += ay*u0[i,:nj-2,k]
    u1[i,j,k] += ay*u0[i,2:,k]
    u1[i,j,k] += az*u0[i,j,:nk-2]
    u1[i,j,k] += az*u0[i,j,2:]
    u1[i,j,k] *= inv_dxyz
    # u1=u0 on boundary so don't need to exclude from norm()
    # Need to ravel so numba will use 1D c-impl of norm()
    return np.linalg.norm(np.ravel(u1 - u0))

@nb.stencil
def poisson_stencil(u, f, ax, ay, az, inv_dxyz):
    return inv_dxyz * (
        ax*u[-1, 0, 0] +
        ax*u[ 1, 0, 0] +
        ay*u[ 0,-1, 0] +
        ay*u[ 0, 1, 0] +
        az*u[ 0, 0,-1] +
        az*u[ 0, 0, 1] -
        f[0, 0, 0]
    )

def stencil_update(u0, u1, f, ax, ay, az, inv_dxyz):
    poisson_stencil(u0, f, ax, ay, az, inv_dxyz, out=u1)
    # u1=u0 on boundary so don't need to exclude from norm()
    # Need to ravel so numba can find 1D c-impl of norm()
    return np.linalg.norm(np.ravel(u1-u0))

def main(argv):

    # Argument Processing
    kn  = argv.pop(0)        if argv else 'vector'
    nx  = int(argv.pop(0))+2 if argv else 18 # +2 for boundary nodes
    ny  = int(argv.pop(0))+2 if argv else nx
    nz  = int(argv.pop(0))+2 if argv else nx
    nt  = int(argv.pop(0))   if argv else 1000
    npr = int(argv.pop(0))   if argv else 100

    # Allocate and initialize solution, source term
    x = np.linspace(0.0, 1.0, nx).reshape(nx,1,1)
    y = np.linspace(0.0, 1.0, ny).reshape(1,ny,1)
    z = np.linspace(0.0, 1.0, nz).reshape(1,1,nz)
    r2 = x*x + y*y + z*z
    u0 = np.exp(-r2)
    f  = (4.0*r2 - 6.0)*u0
    u0[1:nx-1, 1:ny-1, 1:nz-1] = 0.0
    u1 = np.copy(u0)

    # Iterate
    update_kernel = {
        'simple':      simple_update,
        'vector':      vectorized_update,
        'stencil':     stencil_update,
        'simple_jit':  nb.njit(simple_update),
        'vector_jit':  nb.njit(vectorized_update),
        'stencil_jit': nb.njit(stencil_update),
        'simple_par':  nb.njit(simple_update, parallel=True),
        'vector_par':  nb.njit(vectorized_update, parallel=True),
        'stencil_par': nb.njit(stencil_update, parallel=True),
    }
    metrics = Metrics(nx, ny, nz)
    mvalues = (metrics.ax, metrics.ay, metrics.az, metrics.inv_dxyz)
    start = time.perf_counter()
    for n in range(1,nt+1):
        # Pass metrics as scalars so we can jit with Numba
        norm_du = update_kernel[kn](u0, u1, f, *mvalues)
        if n % npr == 0:
            print(f"{n:6d} {norm_du/(nx*ny*nz):12.6e}")
        u0, u1 = u1, u0
    end = time.perf_counter()

    # Extract check value
    u_check = u0[nx//2, ny//2, nz//2]
    u_exact = np.exp(-r2[nx//2, ny//2, nz//2])
    u_error = np.abs(u_check - u_exact)

    # Result summary
    dt = end - start
    mups = (nt/1.0e6)*(nx-2)*(ny-2)*(nz-2)/dt
    print(f"\nCheck Value:       {u_check:8.6f}")
    print(f"Exact Value:       {u_exact:8.6f}")
    print(f"Check Value Error: {u_error:8.6f}")
    print(f"Elasped Time:      {dt:.2f} sec")
    print(f"Throughput:        {mups:.2f} MUPS\n\n")

if __name__ == '__main__':
    main(sys.argv[1:])

File: cpp/poisson.cpp
#include <array>
#include <cassert>
#include <cstdio>
#include <cmath>
#include <string>
#include <map>
#include <vector>
#include "mdspan.hpp"
#include "stopwatch.hpp"

using namespace std::string_literals;
namespace stdex = std::experimental;
using Span3d = stdex::mdspan<double, stdex::extents<
    stdex::dynamic_extent,
    stdex::dynamic_extent,
    stdex::dynamic_extent
>>;
using CSpan3d = stdex::mdspan<const double, stdex::extents<
    stdex::dynamic_extent,
    stdex::dynamic_extent,
    stdex::dynamic_extent
>>;
using Double2 = std::array<double,2>;
using Int3 = std::array<int,3>;
using std::pair;

struct Metrics {
    double dx, dy, dz;
    double ax, ay, az, dxyz;
    double inv_dxyz;
    Metrics(int nx, int ny, int nz) {
        dx = 1.0 / (nx-1);
        dy = 1.0 / (ny-1);
        dz = 1.0 / (nz-1);
        ax = 1.0 / (dx*dx);
        ay = 1.0 / (dy*dy);
        az = 1.0 / (dz*dz);
        dxyz = 2.0 * (ax + ay + az);
        inv_dxyz = 1.0/dxyz;
    }
};

inline Double2 solution(int i, int j, int k, const Metrics& m) {
    auto x  = i*m.dx;
    auto y  = j*m.dy;
    auto z  = k*m.dz;
    auto r2 = x*x + y*y + z*z;
    auto u  = std::exp(-r2);
    auto f  = (4.0*r2 - 6.0)*u;
    return {u, f};
}

inline Double2 stencil(int i, int j, int k, const CSpan3d u0, const CSpan3d f, const Metrics& m) {
    double u = -f(i,j,k);
    u += m.ax*u0(i-1,j,k);
    u += m.ax*u0(i+1,j,k);
    u += m.ay*u0(i,j-1,k);
    u += m.ay*u0(i,j+1,k);
    u += m.az*u0(i,j,k-1);
    u += m.az*u0(i,j,k+1);
    u *= m.inv_dxyz;
    double du = u - u0(i,j,k);
    return {u, du};
}

double simple_update(const CSpan3d u0, const Span3d u1, const CSpan3d f, const Metrics& m) {
    double norm_du = 0.0;
    for (int i = 1; i < u0.extent(0)-1; ++i) {
    for (int j = 1; j < u0.extent(1)-1; ++j) {
    for (int k = 1; k < u0.extent(2)-1; ++k) {
        auto [u, du] = stencil(i, j, k, u0, f, m);
        u1(i,j,k) = u;
        norm_du += du*du;
    }}}
    return std::sqrt(norm_du);
}

double simple_update_omp(const CSpan3d u0, const Span3d u1, const CSpan3d f, const Metrics& m) {
    double norm_du = 0.0;
    #pragma omp parallel for collapse(2) reduction(+:norm_du)
    for (int i = 1; i < u0.extent(0)-1; ++i) {
    for (int j = 1; j < u0.extent(1)-1; ++j) {
    for (int k = 1; k < u0.extent(2)-1; ++k) {
        auto [u, du] = stencil(i, j, k, u0, f, m);
        u1(i,j,k) = u;
        norm_du += du*du;
    }}}
    return std::sqrt(norm_du);
}

double blocked_update(const CSpan3d u0, const Span3d u1, const CSpan3d f, const Metrics& m) {
    constexpr int block_size = 8;
    const int ni = u0.extent(0);
    const int nj = u0.extent(1);
    const int nk = u0.extent(2);
    assert((ni-2) % block_size == 0);
    assert((nj-2) % block_size == 0);
    double norm_du = 0.0;
    for (int ib = 1; ib < ni-1; ib+=block_size) {
    for (int jb = 1; jb < nj-1; jb+=block_size) {
        for (int i = ib; i < ib+block_size; ++i) {
        for (int j = jb; j < jb+block_size; ++j) {
        for (int k = 1;  k < nk-1;          ++k) {
            auto [u, du] = stencil(i, j, k, u0, f, m);
            u1(i,j,k) = u;
            norm_du += du*du;
        }}}
    }}
    return std::sqrt(norm_du);
}

double blocked_update_omp(const CSpan3d u0, const Span3d u1, const CSpan3d f, const Metrics& m) {
    constexpr int block_size = 8;
    const int ni = u0.extent(0);
    const int nj = u0.extent(1);
    const int nk = u0.extent(2);
    assert((ni-2) % block_size == 0);
    assert((nj-2) % block_size == 0);
    double norm_du = 0.0;
    #pragma omp parallel for collapse(2) reduction(+:norm_du)
    for (int ib = 1; ib < ni-1; ib+=block_size) {
    for (int jb = 1; jb < nj-1; jb+=block_size) {
        for (int i = ib; i < ib+block_size; ++i) {
        for (int j = jb; j < jb+block_size; ++j) {
        for (int k = 1;  k < nk-1;          ++k) {
            auto [u, du] = stencil(i, j, k, u0, f, m);
            u1(i,j,k) = u;
            norm_du += du*du;
        }}}
    }}
    return std::sqrt(norm_du);
}

int main(int argc, char* argv[]) {

    // Argument processing
    const std::string kname = (argc > 1)? argv[1] : "simple";
    const int nx = (argc > 2)? std::stoi(argv[2])+2 : 18;  // +2 for boundary nodes
    const int ny = (argc > 3)? std::stoi(argv[3])+2 : nx;
    const int nz = (argc > 4)? std::stoi(argv[4])+2 : nx;
    const int nt = (argc > 5)? std::stoi(argv[5])   : 1000;
    const int np = (argc > 6)? std::stoi(argv[6])   : 100;

    // Select update kernel
    auto kernels = std::map{
        pair{"simple"s,     simple_update},
        pair{"simple_omp"s, simple_update_omp},
        pair{"block"s,      blocked_update},
        pair{"block_omp"s,  blocked_update_omp}
    };
    assert(kernels[kname]);
    auto update_kernel = kernels[kname];

    // Allocate data arrays
    auto storage_u0 = std::vector<double>(nx*ny*nz, 0.0);
    auto storage_u1 = std::vector<double>(nx*ny*nz, 0.0);
    auto storage_f  = std::vector<double>(nx*ny*nz, 0.0);
    auto u0 = Span3d(storage_u0.data(), nx, ny, nz);
    auto u1 = Span3d(storage_u1.data(), nx, ny, nz);
    auto f  = Span3d(storage_f.data(),  nx, ny, nz);

    // Apply boundary conditions, source term
    auto metrics = Metrics(nx,ny,nz);
    for (int i = 0; i < nx; ++i) {
    for (int j = 0; j < ny; ++j) {
    for (int k = 0; k < nz; ++k) {
        const auto [value, source] = solution(i, j, k, metrics);
        bool iboundary = (i == 0) || (i == nx-1);
        bool jboundary = (j == 0) || (j == ny-1);
        bool kboundary = (k == 0) || (k == nz-1);
        if (iboundary || jboundary || kboundary) {
            u0(i,j,k) = value;
            u1(i,j,k) = value;
        }
        f(i,j,k) = source;
    }}}

    // Iterate
    Stopwatch<> sw;
    sw.start();
    for (int n = 1; n <= nt; ++n) {
        auto norm_du = update_kernel(u0, u1, f, metrics);
        if (n % np == 0) printf("%6i %12.6e\n", n, norm_du/(nx*ny*nz));
        std::swap(u1,u0);
    }
    sw.stop();

    // Extract check value
    const auto u_check = u0(nx/2, ny/2, nz/2);
    const auto [u_exact, f_exact] = solution(nx/2, ny/2, nz/2, metrics);

    // Result summary
    auto dt = sw.get_elapsed();
    printf("\nKernel:            %s\n", kname.c_str());
    printf("Check Value:       %8.6f\n", u_check);
    printf("Exact Value:       %8.6f\n", u_exact);
    printf("Check Value Error: %8.6f\n", std::abs(u_check - u_exact));
    printf("Elapsed Time:      %.2f sec\n", dt);
    printf("Throughput:        %.2f MUPS\n\n", (nt/1.0e6)*(nx-2)*(ny-2)*(nz-2)/dt);
    return 0;

}
